src/passes/LoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,246 @@
+      }
+      // If this may branch, we are done.
+      EffectAnalyzer effects(getPassOptions(), curr);
+      if (effects.branches) {
+        break;
+      }
+      if (interestingToMove(curr)) {
+        // Let's see if we can move this out.
+        // Global side effects would prevent this - we might end up
+        // executing them just once.
+        // And we must also move across anything not moved out already,
+        // so check for issues there too.
+        // The rest of the loop's effects matter too, we must also
+        // take into account global state like interacting loads and
+        // stores.
+        if (!effects.hasGlobalSideEffects() &&
+            !effectsSoFar.invalidates(effects) &&
+            !(effects.noticesGlobalSideEffects() && loopEffects.hasGlobalSideEffects())) {
+          // So far so good. Check if our local dependencies are all
+          // outside of the loop, in which case everything is good -
+          // either they are before the loop and constant for us, or
+          // they are after and don't matter.
+          bool canMove = true;
+          if (!effects.localsRead.empty()) {
+            FindAll<GetLocal> gets(curr);
+            for (auto* get : gets.list) {
+              auto& sets = localGraph->getSetses[get];
+              for (auto* set : sets) {
+                // nullptr means a parameter or zero-init value;
+                // no danger to us.
+                if (!set) continue;
+                // Check if the set is in the loop. If not, it's either before,
+                // which is fine, or after, which is also fine - moving curr
+                // to just outside the loop will preserve those relationships.
+                // TODO: this still counts curr's sets as inside the loop, which
+                //       might matter in non-flat mode.
+                if (loopSets.count(set)) {
+                  canMove = false;
+                  break;
+                }
+              }
+              if (!canMove) {
+                break;
+              }
+            }
+          }
+          if (canMove) {
+            // We have checked if our gets are influenced by sets in the loop, and
+            // must also check if our sets interfere with them. To do so, assume
+            // temporarily that we are moving curr out; see if any sets remain for
+            // its indexes.
+            FindAll<SetLocal> currSets(curr);
+            for (auto* set : currSets.list) {
+              assert(numSetsForIndex[set->index] > 0);
+              numSetsForIndex[set->index]--;
+            }
+            for (auto* set : currSets.list) {
+              if (numSetsForIndex[set->index] > 0) {
+                canMove = false;
+                break;
+              }
+            }
+            if (!canMove) {
+              // We failed to move the code, undo those changes.
+              for (auto* set : currSets.list) {
```

Comment:
Hm, rather than undoing the changes, how about having a second list, `updatedNumSetsForIndex`, counting up and seeing if the number of sets is identical. That makes this else clause unnecessary.

Reply:
Might be slightly clearer, I agree, but on the other hand it means initializing an extra array of the size of `numSetsForIndex`, which is of size `numLocals * 4` bytes (could be a few K, and this is for each item we try to move out).

If this pass were larger or more complicated I'd lean more towards clarity over speed, but as it is I slightly lean towards speed.